File: android/src/main/java/com/pspdfkit/flutter/pspdfkit/document/HeadlessDocumentApiImpl.kt
/*
 * Copyright © 2024-2026 PSPDFKit GmbH. All rights reserved.
 * <p>
 * THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
 * AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE PSPDFKIT LICENSE AGREEMENT.
 * UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
 * This notice may not be removed from this file.
 */

package com.pspdfkit.flutter.pspdfkit.document

import android.content.Context
import android.net.Uri
import com.pspdfkit.document.DocumentSource
import com.pspdfkit.document.PdfDocumentLoader
import com.pspdfkit.flutter.pspdfkit.api.HeadlessDocumentApi
import com.pspdfkit.flutter.pspdfkit.api.HeadlessDocumentOpenOptions
import com.pspdfkit.flutter.pspdfkit.api.NutrientApiError
import com.pspdfkit.flutter.pspdfkit.api.PdfDocumentApi
import io.flutter.plugin.common.BinaryMessenger
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers
import java.util.UUID

/**
 * Implementation of HeadlessDocumentApi that provides headless document operations.
 *
 * This implementation allows opening PDF documents without a viewer UI and performing
 * operations like annotation processing, form filling, and document manipulation.
 * Each opened document gets a unique ID and its own PdfDocumentApi channel for
 * isolated communication with Flutter.
 *
 * @param context Android application context for document loading
 * @param binaryMessenger Flutter binary messenger for setting up API channels
 */
class HeadlessDocumentApiImpl(
    private val context: Context,
    private val binaryMessenger: BinaryMessenger
) : HeadlessDocumentApi {

    private var disposable: Disposable? = null

    /**
     * Opens a document from the given path without displaying a viewer.
     *
     * This method loads a PDF document using PdfDocumentLoader and registers it with
     * a unique document ID. A PdfDocumentApi channel is set up for the document using
     * the document ID as a channel suffix, allowing Flutter to communicate with this
     * specific document instance.
     *
     * @param documentPath Path to the PDF document (file path or content:// URI)
     * @param options Optional settings like password for encrypted documents
     * @param callback Callback with Result containing the unique document ID or error
     */
    override fun openDocument(
        documentPath: String,
        options: HeadlessDocumentOpenOptions?,
        callback: (Result<String>) -> Unit
    ) {
        try {
            // Parse the document path as URI
            // If the path is a plain file path (starts with /), convert it to a file:// URI
            val documentUri = if (documentPath.startsWith("/")) {
                Uri.fromFile(java.io.File(documentPath))
            } else {
                Uri.parse(documentPath)
            }

            // Create document source with optional password
            val documentSource = if (options?.password != null) {
                DocumentSource(documentUri, options.password)
            } else {
                DocumentSource(documentUri)
            }

            // Load document asynchronously
            disposable = PdfDocumentLoader.openDocumentAsync(context, documentSource)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { pdfDocument ->
                        // Generate unique document ID
                        val documentId = UUID.randomUUID().toString()

                        // Create FlutterPdfDocument wrapper with documentId and binaryMessenger
                        // for proper cleanup in closeDocument
                        val flutterPdfDocument = FlutterPdfDocument(pdfDocument, documentId, binaryMessenger)

                        // Register document in the global registry
                        FlutterPdfDocument.registerDocument(documentId, flutterPdfDocument)

                        // Setup PdfDocumentApi channel with documentId as suffix
                        PdfDocumentApi.setUp(binaryMessenger, flutterPdfDocument, documentId)

                        // Return document ID to Flutter
                        callback(Result.success(documentId))
                    },
                    { throwable ->
                        callback(
                            Result.failure(
                                NutrientApiError(
                                    "DocumentOpenError",
                                    "Failed to open document: ${throwable.message}",
                                    throwable.stackTraceToString()
                                )
                            )
                        )
                    }
                )
        } catch (e: Exception) {
            callback(
                Result.failure(
                    NutrientApiError(
                        "InvalidDocumentPath",
                        "Invalid document path: ${e.message}",
                        e.stackTraceToString()
                    )
                )
            )
        }
    }

    /**
     * Disposes of all resources held by this API implementation.
     * This should be called when the API is no longer needed.
     */
    fun dispose() {
        disposable?.dispose()
    }
}

File: ios/Classes/HeadlessDocumentApiImpl.swift
//
//  Copyright © 2024-2026 PSPDFKit GmbH. All rights reserved.
//
//  THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
//  AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE PSPDFKIT LICENSE AGREEMENT.
//  UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
//  This notice may not be removed from this file.
//

import Foundation
import PSPDFKit

@objc(HeadlessDocumentApiImpl)
public class HeadlessDocumentApiImpl: NSObject, HeadlessDocumentApi {

    // MARK: - Properties

    /// Registry to store opened documents by their ID
    private static var documentRegistry: [String: FlutterPdfDocument] = [:]

    /// Binary messenger for registering PdfDocumentApi instances
    private var binaryMessenger: FlutterBinaryMessenger

    // MARK: - Initialization

    @objc public init(binaryMessenger: FlutterBinaryMessenger) {
        self.binaryMessenger = binaryMessenger
        super.init()
    }

    // MARK: - HeadlessDocumentApi Implementation

    func openDocument(documentPath: String, options: HeadlessDocumentOpenOptions?, completion: @escaping (Result<String, Error>) -> Void) {
        // Validate document path
        guard !documentPath.isEmpty else {
            let error = NutrientApiError(
                code: "invalid_argument",
                message: "Document path cannot be empty",
                details: nil
            )
            completion(.failure(error))
            return
        }

        // Create document instance
        guard let document = PspdfkitFlutterHelper.document(fromPath: documentPath) else {
            let error = NutrientApiError(
                code: "document_load_error",
                message: "Failed to load document at path: \(documentPath)",
                details: nil
            )
            completion(.failure(error))
            return
        }

        // Handle password-protected documents
        if let password = options?.password {
            PspdfkitFlutterHelper.unlock(document: document, dictionary: ["password": password])
            // Check if the document is still locked after attempting to unlock
            if document.isLocked {
                let error = NutrientApiError(
                    code: "unlock_error",
                    message: "Failed to unlock document with provided password",
                    details: nil
                )
                completion(.failure(error))
                return
            }
        }

        // Validate document is accessible
        guard document.isValid else {
            let error = NutrientApiError(
                code: "document_invalid",
                message: "Document is not valid or accessible",
                details: nil
            )
            completion(.failure(error))
            return
        }

        // Generate unique document ID
        let documentId = UUID().uuidString

        // Create FlutterPdfDocument instance for headless mode with the documentId
        // so closeDocument can properly clean up
        let flutterPdfDocument = FlutterPdfDocument(document: document, documentId: documentId)

        // Register the document in our registry
        HeadlessDocumentApiImpl.documentRegistry[documentId] = flutterPdfDocument

        // Register PdfDocumentApi with the document ID as channel suffix
        // and store the binaryMessenger for cleanup in closeDocument
        flutterPdfDocument.register(binaryMessenger: binaryMessenger)

        // Return the document ID
        completion(.success(documentId))
    }

    // MARK: - Document Registry Management

    /// Closes and removes a document from the registry
    @objc public static func closeDocument(documentId: String) {
        documentRegistry.removeValue(forKey: documentId)
    }

    /// Gets a document from the registry
    @objc public static func getDocument(documentId: String) -> FlutterPdfDocument? {
        return documentRegistry[documentId]
    }

    /// Clears all documents from the registry
    @objc public static func clearAllDocuments() {
        documentRegistry.removeAll()
    }

    // MARK: - Registration

    @objc public func register(binaryMessenger: FlutterBinaryMessenger) {
        HeadlessDocumentApiSetup.setUp(binaryMessenger: binaryMessenger, api: self, messageChannelSuffix: "nutrient")
    }

    @objc public func unregister(binaryMessenger: FlutterBinaryMessenger) {
        HeadlessDocumentApiSetup.setUp(binaryMessenger: binaryMessenger, api: nil, messageChannelSuffix: "nutrient")
    }
}
